docs/adr/000x-modelling-history.md

```diff
@@ -0,0 +1,314 @@
+updated to the most recent foreign keys which makes them not suitable vessels
+for storing data that may change.
+
+The choice of this immutability strategy is to keep the concerns of the present
+and history stored in the same data stores - thus ensuring history remains a
+first class citizen. A nice side effect of having immutable models is
+this opens options for caching - since we know that data for that
+model will never change it can effectively be cached forever.
+
+### Breakdown of Revision
+
+Revision is a model that stores little data itself and joins to other models
+that store the data. This can be visualised as:
+
+![Revision breakdown](000x/revision-diagram.png)
+
+The intention of breaking this up is to be conservative with the amount of data
+duplicated for both storage and maintenance reasons. Thus when a user edits
+the title of an edition a new ContentRevision object is created and the other
+objects (TagsRevision, MetadataRevision and ImageRevisions) are associated with
+the next revision.
+
+It is intended that delegation be used when interfacing with a revision to
+enable accessing data within the child objects of a revision without the caller
+being directly aware of them.
+
+### Image modelling
+
+Images are modelled in a similar pattern to Revision. At their core is a
+concept of an Image::Revision which stores joins to other models, as
+represented below:
+
+![Image Revision breakdown](000x/image-revision-diagram.png)
+
+The Image model itself is used for continuation between image revisions, we know
+that two Image::Revisions are versions of the same item if they share the same
+Image association.
+
+The data of an Image::Revision is stored between an Image::FileRevision and an
+Image::MetadataRevision. These two objects differ by the fact that any change to
+Image::FileRevision requires changes to the resultant files (such as crop
+dimensions), whereas Image::MetadataRevision stores accompanying data that
+doesn't affect the files (such as alt text).
+
+Image::FileRevision is associated with an ActiveStorage::Blob object that is
+responsible for managing the storage of the source file. It also has a one to
+many association with Image::Asset, each of these represent resulting files that
+are uploaded to Asset Manager. For a single image uploaded we'd likely upload
+multiple sizes to Asset Manager - these different files are named variants. An
+Image::Asset object stores whether the asset is on Asset Manager, what state it
+is in and the URL.
+
+### Timeline entry
+
+A timeline entry represents an event that should be shown to a user as part
+of a documents history. This stores the event which occurred and associations
+to relevant models. The sole purpose of this is for representing a
+timeline to users visually and not for storing historical metadata.
+
+The TimelineEntry model should not store data which could not be
+derived from other aspects of a document. This is to maintain
```

Comment:
`The TimelineEntry model should not store data which could not be derived from other aspects of a document` -  could you give an example of what data this is referring to?

Reply:
We had a previous example of this where Removal, Retire and InternalNote were associated to a TimelineEntry rather than to a document directly.

This meant that when we wanted to show that a document was removed we had to query timeline entry to get the first removal. So in this instance TimelineEntry was being queried to find out what the removal was for an Edition

Reply:
I've included that information in the doc now.

Reply:
Nice!